### PR TITLE
HHH-10000

### PR DESCRIPTION
**Pull request by drepshas**
As Teradata requires that primary keys be non-null, changes were made to Dialect.java and ModelBinder.java to accommodate this. The changes to ModelBinder also imply that the hibernate.check_nullability property be set to false when testing. So the changes required to: hibernate-core/src/test/resources/hibernate.properties for Teradata testing are:
hibernate.check_nullability=false
hibernate.connection.autocommit=true

The change to DriverManagerConnectionProviderImpl.java was done to rollback any pending transactions when a connection was returned to the pool. What I saw when testing was that a lot of hang problems were due to an earlier failure on a specific connection that was in the middle of a transaction.

### CODE REVIEW

**Comment by sebersole** [comment]
The mainline hibernate-core/src/test/resources/hibernate.properties is not the correct place to make changes for the testing of a particular Dialect.  The proper way to do that is via the database matrix plugin, defining a new testing database profile for that dialect.

**Comment by gbadner** [comment]
There are lots of changes to mappings to satisfy Teradata's requirements. I am not in favor of making those changes because we want to test valid functionality of the other dialects.

If teradata has special requirements, then:
- skip the tests for teradata if it has a mapping that doesn't work for teradata;
- create a new test with mapping specifically for teradata.

File: hibernate-core/src/main/java/org/hibernate/boot/model/source/internal/hbm/ModelBinder.java [hunk 1]
```diff
@@ -281,6 +282,20 @@ private void bindRootEntity(EntityHierarchySourceImpl hierarchySource, RootClass
 				rootEntityDescriptor
 		);
 
+		if (database.getDialect().isNonNullPrimaryKeyRequired()) {
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/boot/model/source/internal/hbm/ModelBinder.java]: I wonder if a better option is a "complete()" method on Database where we can then iterate all tables and do this for all pks

**Comment by sebersole** [comment]
We gave some specific feedback, but here is some general feedback.  

First you group of lot of different changes into one Jira and one PR here.  Each unique change should be its own Jira.  Ideally you'd have one PR for each Jira.  Really I don't care if you want to want to group multiple Jiras into a single PR as long as you talk to us first to make sure we agree to all the changes.  Thats really the point of PRs, for discussion and consensus.  

Secondly, the PR is based on a very outdated point on master.  The "bulk id strategy" work shows that.  You have the state of that support as it was before I did major rework to that code 4 months ago.  And who knows how long prior to that the PR is actually based on.  If this were a PR that happened to sit for 4 months, that is a different discussion.  But its not.  This is a new PR.  This work needs to be rebased on the current upstream, especially considering you have changes in code that is significantly different between.

**Pull request closed without merging**